### PR TITLE
 Replace `ConstFnMutClosure` with const closures 

### PR DESCRIPTION
**Pull request by fee1-dead**
Also fixes a parser bug. cc @oli-obk for compiler changes

### CODE REVIEW

File: library/core/src/option.rs [hunk 1]
```diff
@@ -1969,7 +1969,8 @@ impl<T> const Default for Option<T> {
 }
 
 #[stable(feature = "rust1", since = "1.0.0")]
-impl<T> IntoIterator for Option<T> {
+#[rustc_const_unstable(feature = "const_iter", issue = "92476")]
+impl<T> const IntoIterator for Option<T> {
```
Comment (chriss0612) [inline library/core/src/option.rs:1973]: Why is this included in this PR?

Comment (fee1-dead) [reply library/core/src/option.rs]: Fixed, thanks.
